qiskit/circuit/quantumcircuitdata.py

```diff
@@ -59,52 +133,71 @@ def __setitem__(self, key, value):
         self._circuit._update_parameter_table(instruction)
 
     def insert(self, index, value):
-        self._circuit._data.insert(index, None)
-        self[index] = value
-
-    def __delitem__(self, i):
-        del self._circuit._data[i]
+        try:
+            # To consolidate validation to __setitem__, insert None and overwrite.
+            self._list.insert(index, None)
+            self[index] = value
+        except CircuitError as err:
+            del self._list[index]
+            raise err
```

Comment:
Alternatively, would it be cleaner to factor out the validation into a separate, private instance method, and then just have `__setitem__` and `insert` both call it?

Reply:
Thinking about this a bit more, I think yes, but this would require a not inconsiderable refactor to both the circuit and register. We're lumping a few concerns behind the concept of "validation" (is the item being added a valid candidate member for this data structure, and following this addition/modification, is the data structure left in a consistent state) which are currently implemented throughout `QuantumCircuit.add_register`, `QuantumCircuit.add_bits`, and `Register.__init__`. If keeping these list-like interfaces is something we intend to support in the long term (and likely even if not), I agree there are definite opportunities for re-structuring to make this logic cleaner which I think we can tackle as a separate issue.